audit/audit.go

```diff
@@ -0,0 +1,272 @@
+	client *base.Client
+	// replaceClient is used to ensure we only replace the client once. This is needed because we can have multiple
+	// goroutines trying to replace the client at the same time.
+	replacingClient atomic.Bool
+
+	// create is the function that creates a new connection to the remote audit server. This is used when the
+	// connection is broken.
+	create CreateConn
+	// backoff is the exponential backoff for making a new connection to the remote audit server.
+	backoff *exponential.Backoff
+	policy  exponential.Policy
+
+	// notifier is the channel that will receive errors when the connection to the remote audit server is broken or
+	// the queue is full. If this channel is full, the error will be dropped.
+	notifier chan NotifyError
+	// auditOptions are the options for the underlying base.Client.
+	auditOptions []base.Option
+
+	// replaceBackoffRunner is a function that replaces the audit client with a new one using an exponential backoff.
+	// In production, this uses .replaceBackoff. In testing this can be replaced with a function that does nothing.
+	replaceBackoffRunner func(ctx context.Context) error
+	// sendRunner is a function that sends a message to the remote audit server.
+	// In production, this uses .client.Sent(). In testing this can be replaced with a function that does nothing.
+	sendRunner func(ctx context.Context, msg msgs.Msg) error
+}
+
+// CreateConn is a function that creates a connection to a remote audit server.
+type CreateConn func() (conn.Audit, error)
+
+// Option is an option for the smart client.
+type Option func(*Client) error
+
+// WithAuditOptions sets the audit options for the underlying audit client.
+func WithAuditOptions(options ...base.Option) Option {
+	return func(c *Client) error {
+		c.auditOptions = options
+		return nil
+	}
+}
+
+// WithBackoffPolicy sets the exponential backoff policy for making a new connection to the remote audit server.
+// By default this uses the default backoff settings from the exponential package.
+func WithExponentialBackoff(policy exponential.Policy) Option {
+	return func(c *Client) error {
+		c.policy = policy
+		return nil
+	}
+}
+
+// New creates a new smart client to the remote audit server. You can get a CreateConn
+// from conn.NewDomainSocket(), conn.NewTCP() or conn.NewNoop(). The last one is useful for testing
+// when you don't want to send logs anywhere.
+func New(create CreateConn, options ...Option) (*Client, error) {
+	connect, err := create()
+	if err != nil {
+		return nil, err
+	}
+
+	if runtime.GOOS != "linux" && !testing.Testing() {
+		if connect.Type() != conn.TypeNoOP {
+			return nil, fmt.Errorf("audit: only linux is clients can use Audit with non-NoOp connections")
+		}
+	}
+
+	c := &Client{
+		create:   create,
+		notifier: make(chan NotifyError, 1),
+	}
+	c.replaceBackoffRunner = c.replaceBackoff
+	c.replacingClient.Store(false)
+
+	for _, o := range options {
+		o(c)
+	}
+
+	if reflect.ValueOf(c.policy).IsZero() {
```

Comment:
Dodging the non-nil interface w/ nil value? Is this just being careful or is there a specific case where this happened to you already? IIRC the default `exponential.Policy` should `== nil` the issue will only arise when a `nil` of a different "type" is returned and implicitly wrapped?

Reply:
Not at all.

I like to avoid the use of pointers in any place where I don't need them.  They are overused and create an overall burden on the GC, which is what makes Go slow.  Death by 1000 cuts. 

So the policies in the backoff lib are all just `struct` and not `*struct`.

Most Go programmers just use pointers, but they also leak into the GC all the time. Then they wonder why the their large service is slow.  Well, they have *bool and *int...

I could have also done it this way, but I like reflect better and it doesn't leak here:
https://go.dev/play/p/ffAGrRDMJ0H